guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+        password: 'password',
+      });
+      keyringController.cacheEncryptionKey = true;
+      const returnValue = await keyringController.encryptor.decryptWithKey();
+      const decryptWithKeyStub = sinon.stub(
+        keyringController.encryptor,
+        'decryptWithKey',
+      );
+      decryptWithKeyStub.resolves(Promise.resolve(returnValue));
+
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+
+      await keyringController.setLocked();
+
+      await keyringController.submitEncryptionKey(
+        'mockEncryptionKey',
+        'mockEncryptionSalt',
+      );
+
+      expect(keyringController.encryptor.decryptWithKey.calledOnce).toBe(true);
+      expect(keyringController.keyrings).toHaveLength(1);
+    });
+  })
+});
+```
+
+2️⃣
+
+``` typescript
+describe('KeyringController', () => {
+  describe('cacheEncryptionKey', () => {
+    it('unlocks the keyrings with valid information', async () => {
+      const keyringController = await initializeKeyringController({
+        password: 'password',
+      });
+      keyringController.cacheEncryptionKey = true;
+      const returnValue = await keyringController.encryptor.decryptWithKey();
+      const decryptWithKeyStub = sinon.stub(
+        keyringController.encryptor,
+        'decryptWithKey',
+      );
+      decryptWithKeyStub.resolves(Promise.resolve(returnValue));
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+      await keyringController.setLocked();
+
+      await keyringController.submitEncryptionKey(
+        'mockEncryptionKey',
+        'mockEncryptionSalt',
+      );
+
+      expect(keyringController.encryptor.decryptWithKey.calledOnce).toBe(true);
+      expect(keyringController.keyrings).toHaveLength(1);
+    });
+  })
+});
+```
+
+### Read more
+
+- ["Arrange, act, assert"](http://xp123.com/articles/3a-arrange-act-assert/) is a simplification of the four-phase test.
+- In [behavior-driven development](https://www.agilealliance.org/glossary/bdd/), this is also called ["given, when, then"](https://martinfowler.com/bliki/GivenWhenThen.html).
+
+## Keep tests isolated
+
+In order to be useful, tests must be deterministic. That is, a single test within a file must always pass each and every time, whether it is run on its own, alongside other tests in the same file, or other tests across other files.
+
+To achieve determinism, tests must be performed in a clean room. If a test makes changes to any kind of state which persists outside of itself, it must revert those changes after completion to prevent contaminating other tests.
+
+There are a few different forms that this state can take:
+
+### Module mocks
+
+Mocks for modules — objects reachable via `import` lines — are created via `jest.mock` declarations within a test file. Since modules are mocked once when a test file is read, they are global in the context of the file, so changes in one test can "bleed" over into another test:
```

Comment:
Nit: Seems a bit unnecessary to define "module" here - hopefully that's a commonly understood word (especially given your examples). Also it doesn't seem correct to call the `jest.mock` calls "declarations" - nothing is being declared when we mock a module.
```suggestion
Mocks for modules are created by calling `jest.mock` within a test file. Since modules are mocked once when a test file is read, they are global in the context of the file, so changes in one test can "bleed" over into another test:
```

Reply:
I ended up removing this section.